src/java/org/apache/cassandra/utils/concurrent/Semaphore.java

```diff
@@ -0,0 +1,333 @@
+     * @return true iff permits acquired
+     * @throws InterruptedException if interrupted
+     */
+    boolean tryAcquire(int acquire, long time, TimeUnit unit) throws InterruptedException;
+
+    /**
+     * Try to take permits, waiting until the deadline
+     * @return true iff permits acquired
+     * @throws InterruptedException if interrupted
+     */
+    boolean tryAcquireUntil(int acquire, long nanoTimeDeadline) throws InterruptedException;
+
+    /**
+     * Take permits, waiting indefinitely until available
+     * @throws InterruptedException if interrupted
+     */
+    void acquire(int acquire) throws InterruptedException;
+
+    /**
+     * Take permits, waiting indefinitely until available
+     * @throws UncheckedInterruptedException if interrupted
+     */
+    void acquireThrowUncheckedOnInterrupt(int acquire) throws UncheckedInterruptedException;
+
+    /**
+     * Factory method used to capture and redirect instantiations for simulation
+     *
+     * Construct an unfair Semaphore initially holding the specified number of permits
+     */
+    @Intercept
+    public static Semaphore newSemaphore(int permits)
+    {
+        return new UnfairAsync(permits);
+    }
+
+    /**
+     * Factory method used to capture and redirect instantiations for simulation
+     *
+     * Construct a fair Semaphore initially holding the specified number of permits
+     */
+    @Intercept
+    public static Semaphore newFairSemaphore(int permits)
+    {
+        return new FairJDK(permits);
+    }
+
+    /**
+     * An unfair semaphore, making no guarantees about thread starvation.
+     *
+     * TODO this Semaphore is potentially inefficient if used with release quantities other than 1
+     *      (this is unimportant at time of authoring)
+     */
+    public static class UnfairAsync implements Semaphore
+    {
+        private static final AtomicReferenceFieldUpdater<UnfairAsync, WaitQueue> waitingUpdater = AtomicReferenceFieldUpdater.newUpdater(UnfairAsync.class, WaitQueue.class, "waiting");
+        private static final AtomicIntegerFieldUpdater<UnfairAsync> permitsUpdater = AtomicIntegerFieldUpdater.newUpdater(UnfairAsync.class, "permits");
+        private volatile WaitQueue waiting;
+        private volatile int permits;
+
+        // WARNING: if extending this class, consider simulator interactions
+        public UnfairAsync(int permits)
+        {
+            this.permits = permits;
+        }
+
+        /**
+         * {@link Semaphore#drain()}
+         */
+        public int drain()
+        {
+            return permitsUpdater.getAndSet(this, 0);
+        }
+
+        /**
+         * {@link Semaphore#permits()}
+         */
+        public int permits()
+        {
+            return permits;
+        }
+
+        /**
+         * {@link Semaphore#release(int)}
+         */
+        public void release(int permits)
+        {
+            if (permits < 0) throw new IllegalArgumentException();
```

Comment:
Let's add a message.

Reply:
I do not believe adding messages to every IllegalStateException improves the codebase.

Reply:
IMO it helps operators when they see a error message in logs instead of/along with a trace. However, probably in this case you're right, message won't help them to understand the cause.